test/data/BinaryHeap.ts

```diff
@@ -0,0 +1,495 @@
+
+          await instance['add(bytes32)'](twoBytes32);
+          expect(await instance['length()']()).to.equal(3);
+
+          await instance['remove(bytes32)'](twoBytes32);
+          expect(await instance['length()']()).to.equal(2);
+
+          await instance['remove(bytes32)'](oneBytes32);
+          expect(await instance['length()']()).to.equal(1);
+
+          await instance['remove(bytes32)'](zeroBytes32);
+          expect(await instance['length()']()).to.equal(0);
+        });
+      });
+
+      describe('#root()', () => {
+        it('returns value of the root node', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          expect(await instance['root()']()).to.equal(zeroBytes32);
+
+          await instance['add(bytes32)'](oneBytes32);
+          expect(await instance['root()']()).to.equal(oneBytes32);
+
+          await instance['add(bytes32)'](twoBytes32);
+          expect(await instance['root()']()).to.equal(twoBytes32);
+
+          await instance['remove(bytes32)'](twoBytes32);
+          expect(await instance['root()']()).to.equal(oneBytes32);
+
+          await instance['remove(bytes32)'](oneBytes32);
+          expect(await instance['root()']()).to.equal(zeroBytes32);
+        });
+
+        describe('reverts if', function () {
+          it('index out of bounds', async () => {
+            await expect(instance['root()']()).to.be.revertedWith(
+              'EnumerableHeap: index out of bounds',
+            );
+          });
+        });
+      });
+
+      describe('#add(bytes32)', () => {
+        it('adds value to heap', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          expect(await instance['heap()']()).to.deep.equal([
+            twoBytes32,
+            zeroBytes32,
+            oneBytes32,
+          ]);
+        });
+      });
+
+      describe('#remove(bytes32)', () => {
+        it('removes value from heap', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          await instance['remove(bytes32)'](zeroBytes32);
+          expect(await instance['heap()']()).to.deep.equal([
+            twoBytes32,
+            oneBytes32,
+          ]);
+        });
+
+        it('returns true if value is removed', async () => {
+          await instance['add(bytes32)'](zeroBytes32);
+          await instance['add(bytes32)'](oneBytes32);
+          await instance['add(bytes32)'](twoBytes32);
+
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.true;
+        });
+
+        it('returns false if value is not removed', async () => {
+          expect(await instance.callStatic['remove(bytes32)'](zeroBytes32)).to
+            .be.false;
+        });
+      });
+    });
```

Comment:
Perhaps dd explicit test for `_swap` and `_maxHeapify`?

Reply:
@NouDaimon I think you've touched on something that I haven't seen so far in SS. `_swap` and `_maxHeapify` are both private functions that are not exposed externally like the other functions in the `BinaryHeap` library. This presents a unique challenge for writing tests. I think there are three possible solutions here, keep in mind, this situation may come up again so this could be an excellent time to have a plan for dealing with it.

1. Keep `_swap` and `_maxHeapify` private but copy-paste their logic to the mock binary heap files.
2. Set `_swap` and `_maxHeapify` as internal functions
3. Don't test `_swap` and `_maxHeapify` directly

Would certainly be interested in hearing your thoughts on this. I'm torn between 2 and 3, on one hand, I think it makes sense to test as much as possible but adding these functions to the interface doesn't sit well with me. On the other hand, a robust set of tests may make it unnecessary to test each and every function, but this makes it challenging to track regressions and cover edge cases.

cc @ItsNickBarry

Reply:
I vote for 3, in the cases of `private` functions - `internal` functions tend to have functions exposing them and can be tested in that manner.

Every function that invokes `private` functions should test that the `private` functions work as expected, since they are now part of the function being tested. As a result, the test for each function invoking `private` functions has to be more robust - this will aid in catching edge cases as well, should the `private` functionality change.

The goal with the above practice is to make tests as self-contained as possible.

Using the `BinaryHeap` as an example, the tests for `_remove` should check that the root of the remaining heap is indeed node with the largest (since `_maxHeapify`) value - this ensure that this specific result of the `_maxHeapify` function which is invoked in `_remove` works as expected.

Reply:
No need to test `private` functions.  All behavior of such functions should be testable via the `internal` or `external` functions.